### PR TITLE
Shading Guava.

### PR DESCRIPTION
**Pull request by oza**
This PR tries to solve some problems described in #46 and #16. grpc and protobuf can also be shaded here.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -65,6 +73,12 @@
             <artifactId>slf4j-api</artifactId>
             <version>${slf4j.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>${guava.version}</version>
```
Comment (sijie) [inline pom.xml]: since gRpc depends on guava, I think you don't have to list the exact version here, no?

**Comment by sijie** [comment]
@oza I am actually not sure shading a dependency for a public interface is right solution. for example, if my project uses guava ListenableFuture and etcd exposes a re-packaged ListenableFuture. It will become a bit weird because they are all *guava* ListenableFuture. However, I can't simply chain the callbacks, I have to write another adapter to covert ectd re-packaged ListenableFuture to guava ListenableFuture.

I would suggest not shade in etcd. If the application has dependency conflicts with the dependencies that jetcd has, it won't be just simply guava version conflicts. There will be other dependency conflicts. so this should really leave application to resolve the conflicts rather than the library itself.

**Comment by oza** [comment]
@sijie @xiang90 make sense. IMO, it's better to avoid shading since we have iPOJO abstraction and we can easily avoid conflict.Shading can cause a problem to confuse users because of same name of classes, to make fat binary.

@lburgazzoli what do you think? Fortunately, jetcd has not yet decided the interfaces. Hence, we can change the interfaces instead of shading.

**Comment by sijie** [comment]
I would say lets use the current ListenableFuture to complete the first version of jetcd. When we approach the first release version for jetcd, lets decide what is the best *Future*/*Callback* interface. does that make sense to everyone here?

**Comment by lburgazzoli** [comment]
Shading guava would make sense if guava classes are not exposed to public api. 

I would say that such decision of how interfaces should look like is something to address as early as possible to avoid breaking changes after release.

**Comment by oza** [comment]
> I would say lets use the current ListenableFuture to complete the first version of jetcd. When we approach the first release version for jetcd, lets decide what is the best Future/Callback interface. does that make sense to everyone here?

As the same reason that we shouldn't expose ByteString of protobuf to end-user [here](https://github.com/coreos/jetcd/pull/45#discussion_r76964462), I think it's not good option. It's OK to use ListenableFuture internally. However, if we mix bytecode of different version Guava, it can cause method missing error. I also breaks iPOJO class path isolation by Apache Felix. To avoid them, we can have a 2 choices:

1. Changing interface.
2. Shading Guava without changing interface.

Please correct me if I'm wrong.

**Comment by sijie** [comment]
I am okay with changing to java Future and byte[] or ByteBuffer, which are those standard libs.

**Comment by lburgazzoli** [comment]
Wouldn't CompletableFuture be better ?

**Comment by oza** [comment]
@lburgazzoli CompletableFuture seems to be better to me.

**Comment by AdoHe** [comment]
Agree. I think it's time for use to make a decision, I would support use CompletableFuture and custom ByteSequence.

**Comment by sijie** [comment]
CompletableFuture works for me. why not java nio ByteBuffer?

**Comment by oza** [comment]
@sijie Good catch. I would like to suggest that ByteSequence accept both ByteBuffer and byte[]. It's because that both of ByteBuffer and byte[] are used for buffer management(e.g. protobuf's accepts both of them).

I think ByteBuffer is used for safe buffer management by switching read/write mode, and byte[] is used for its simplicity.

**Comment by fanminshi** [comment]
I replace usage of ByteString from all exposed API to ByteSequence in https://github.com/coreos/jetcd/pull/47. I think we don't shading guava library anymore since jetcd APIs doesn't depends on guava.

**Pull request closed without merging**